lib/pos_network.go

```diff
@@ -400,3 +400,129 @@ func DecodeAggregatedBLSSignature(rr io.Reader) (*AggregatedBLSSignature, error)
 
 	return &sig, nil
 }
+
+// TimeoutAggregateQuorumCertificate is an aggregation of timeout messages from 2/3rds
+// of all validators, weighted by stake, which indicates that these validators want to
+// time out a particular view.
+//
+// When validators want to time out a view, they send their high QCs to the block proposer
+// who builds an aggregate QC extending the chain from the highest QC that it received
+// from all validators who timed out. To prove that it has selected the highest QC, the
+// proposer also includes a list of the high QC views that each validator has sent.
+type TimeoutAggregateQuorumCertificate struct {
+
+	// The view that the block proposers has produced a timeout QC for.
+	TimedOutView uint64
+
+	// This is the highest QC that the block proposer received from any validator who
+	// has timed out for the current view.
+	ValidatorsHighQC *QuorumCertificate
+
+	// Here we include a list of the HighQC.View values we got from each of the
+	// validators in the ValidatorsTimeoutHighQCViews field. In addition, for each
+	// unique HighQC.View value we received, we combine all the partial signatures
+	// for that HighQC.View into a single BLSMultiSignature.
+	//
+	// Note: The aggregated signature is made up of partial signatures for all
+	// present validators, each of whom signed a payload with the pair
+	// (current view, the validator's local HighQC.View).
+	ValidatorsTimeoutHighQCViews         []uint64
+	ValidatorsTimeoutAggregatedSignature *AggregatedBLSSignature
+}
+
+// Performs a deep equality check between two TimeoutAggregateQuorumCertificate, and
+// returns true if the values of the two are identical.
+func (aggQC *TimeoutAggregateQuorumCertificate) Eq(
```

Comment:
This is a nice in-depth deep comparison. But there is also a dumber method: encode both objects to bytes and see if they are equal. Assuming encoding is bijective that should be chill. The current code is fine to keep, but maybe in the future, the dumb comparison will save u time.

Reply:
Ah this is clever. It shouldn't be possible for the same byte encoded object to decode into two different struct values, so this will work. Thanks for the suggestion.

Reply:
Ah, this would be nice! The one downside is that `aggQC .ToBytes()` may return an error if the aggQC has nil fields that cannot serialize to bytes. In that case, two otherwise equivalent `TimeoutAggregateQuorumCertificate` structs values may end up with an undefined equality result. We need to keep the existing field by field deep comparison.